source/common/http/filter/ext_authz.cc

```diff
@@ -0,0 +1,132 @@
+#include "common/http/filter/ext_authz.h"
+
+#include <string>
+#include <vector>
+
+#include "envoy/http/codes.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "common/ext_authz/ext_authz_impl.h"
+#include "common/http/codes.h"
+#include "common/router/config_impl.h"
+
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace Http {
+namespace ExtAuthz {
+
+namespace {
+
+static const Http::HeaderMap* getDeniedHeader() {
+  static const Http::HeaderMap* header_map = new Http::HeaderMapImpl{
+      {Http::Headers::get().Status, std::to_string(enumToInt(Code::Forbidden))}};
+  return header_map;
+}
+
+} // namespace
+
+void Filter::initiateCall(const HeaderMap& headers) {
+
+  Router::RouteConstSharedPtr route = callbacks_->route();
+  if (!route || !route->routeEntry()) {
+    return;
+  }
+
+  const Router::RouteEntry* route_entry = route->routeEntry();
+  Upstream::ThreadLocalCluster* cluster = config_->cm().get(route_entry->clusterName());
+  if (!cluster) {
+    return;
+  }
+  cluster_ = cluster->info();
+
+  Envoy::ExtAuthz::CheckRequestUtils::createHttpCheck(callbacks_, headers, checkRequest_);
+
+  state_ = State::Calling;
+  initiating_call_ = true;
+  client_->check(*this, checkRequest_, callbacks_->activeSpan());
+  initiating_call_ = false;
+}
+
+FilterHeadersStatus Filter::decodeHeaders(HeaderMap& headers, bool) {
+  initiateCall(headers);
+  return (state_ == State::Calling || state_ == State::Responded)
```

Comment:
How can state_ == Responded?  You assert in decodeData() that this can't be true.